substrate/primitives/core/src/bandersnatch.rs

```diff
@@ -368,21 +307,21 @@ pub mod vrf {
 
 	#[cfg(feature = "full_crypto")]
 	impl VrfSecret for Pair {
-		fn vrf_sign(&self, data: &Self::VrfSignData) -> Self::VrfSignature {
-			const _: () = assert!(MAX_VRF_IOS == 3, "`MAX_VRF_IOS` expected to be 3");
-			// Workaround to overcome backend signature generic over the number of IOs.
-			match data.inputs.len() {
-				0 => self.vrf_sign_gen::<0>(data),
-				1 => self.vrf_sign_gen::<1>(data),
-				2 => self.vrf_sign_gen::<2>(data),
-				3 => self.vrf_sign_gen::<3>(data),
-				_ => unreachable!(),
-			}
+		fn vrf_sign(&self, data: &VrfSignData) -> VrfSignature {
```

Comment:
Why isn't this implemented in the library? if vrf_sign means to give the pre_output and the proof, it makes sense that a vrf library implement that and simply get called here, instead of people going around to look for the logic if they want to know how it is implemented.

Reply:
The preout computation is left to the user as he may want to check what is the vrf_output (i.e. hash of preout) before doing the signature. E.g. in sassafras we compute the ticket value using the vrf output, then we check if it is below a threshold and finaly we sign if is a valid ticket.

BTW maybe I can provide a wrapper in `ark-vrf` to also compute the preout internally for convenience

Reply:
Yeah, that last hash `real_vrf_out = h(vrf_input,vrf_preout)` always sucks, but randomness requires it.  I always jumped through hoops to hide it, but that's ugly too.

Although safrole removed the multiple VRF input-output pairs sometimes, you do need multiple VRF input-output pairs elsewhere like games, so you kinda need to expose some detached way of handling things. 

I worked out a somewhat sensible trait set in https://github.com/w3f/ring-vrf/blob/master/dleq_vrf/src/traits.rs but it's basically a bunch of (in)convenience functions.